nav2_map_server/src/map_saver.cpp

```diff
@@ -30,221 +31,173 @@
-            } else if (threshold_occupied_ <= map_cell) {
-              pixel = Magick::ColorGray(0 / 255.0);
-            } else {
-              pixel = Magick::ColorGray(205 / 255.0);
-            }
-            break;
-          case MapMode::Scale:
-            if (map_cell < 0 || 100 < map_cell) {
-              pixel = Magick::ColorGray{0.5};
-              pixel.alphaQuantum(TransparentOpacity);
-            } else {
-              pixel = Magick::ColorGray{(100.0 - map_cell) / 100.0};
-            }
-            break;
-          case MapMode::Raw:
-            Magick::Quantum q;
-            if (map_cell < 0 || 100 < map_cell) {
-              q = MaxRGB;
-            } else {
-              q = map_cell / 255.0 * MaxRGB;
-            }
-            pixel = Magick::Color(q, q, q);
-            break;
-          default:
-            throw std::runtime_error("Invalid map mode");
-        }
-        image.pixelColor(x, y, pixel);
+    rclcpp::Time start_time = now();
+    while (rclcpp::ok()) {
+      rclcpp::spin_some(map_listener_);
+
+      if ((now() - start_time) > *save_map_timeout_) {
+        RCLCPP_ERROR(get_logger(), "Failed to save the map: timeout");
+        return false;
       }
-    }
 
-    RCLCPP_INFO(logger, "Writing map occupancy data to %s", mapdatafile.c_str());
-    image.write(mapdatafile);
-  }
+      if (got_map_msg_) {
+        // Map message received. Saving it to file
+        if (saveMapToFile(*msg_, save_parameters)) {
+          RCLCPP_INFO(get_logger(), "Map saved successfully");
+          return true;
+        } else {
+          RCLCPP_ERROR(get_logger(), "Failed to save the map");
+          return false;
+        }
+      }
 
-  std::string mapmetadatafile = mapname_ + ".yaml";
-  {
-    std::ofstream yaml(mapmetadatafile);
-
-    geometry_msgs::msg::Quaternion orientation = map.info.origin.orientation;
-    tf2::Matrix3x3 mat(tf2::Quaternion(orientation.x, orientation.y, orientation.z, orientation.w));
-    double yaw, pitch, roll;
-    mat.getEulerYPR(yaw, pitch, roll);
-
-    YAML::Emitter e;
-    e << YAML::Precision(3);
-    e << YAML::BeginMap;
-    e << YAML::Key << "image" << YAML::Value << mapdatafile;
-    e << YAML::Key << "mode" << YAML::Value << map_mode_to_string(map_mode);
-    e << YAML::Key << "resolution" << YAML::Value << map.info.resolution;
-    e << YAML::Key << "origin" << YAML::Flow << YAML::BeginSeq << map.info.origin.position.x <<
-      map.info.origin.position.y << yaw << YAML::EndSeq;
-    e << YAML::Key << "negate" << YAML::Value << 0;
-    e << YAML::Key << "occupied_thresh" << YAML::Value << threshold_occupied_ / 100.0;
-    e << YAML::Key << "free_thresh" << YAML::Value << threshold_free_ / 100.0;
-
-    if (!e.good()) {
-      RCLCPP_WARN(
-        logger, "YAML writer failed with an error %s. The map metadata may be invalid.",
-        e.GetLastError().c_str());
+      rclcpp::sleep_for(std::chrono::milliseconds(100));
```

Comment:
Thank you for  throttling this.

Reply:
I've made it to increase CPU time in favor of other ROS2 processes (incl. those running the `mapCallback()`). Do you mean, that it is better to remove the sleep (why)?